hazelcast/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlanBuilder.java

```diff
@@ -69,25 +72,49 @@ private ExecutionPlanBuilder() {
 
     @SuppressWarnings({"checkstyle:ParameterNumber", "rawtypes"})
     public static CompletableFuture<Map<MemberInfo, ExecutionPlan>> createExecutionPlans(
-            NodeEngineImpl nodeEngine, List<MemberInfo> memberInfos, DAG dag, long jobId, long executionId,
-            JobConfig jobConfig, long lastSnapshotId, boolean isLightJob, Subject subject
+            NodeEngineImpl nodeEngine,
+            List<MemberInfo> memberInfos,
+            DAG dag,
+            long jobId,
+            long executionId,
+            JobConfig jobConfig,
+            long lastSnapshotId,
+            boolean isLightJob,
+            Subject subject
     ) {
         final VerticesIdAndOrder verticesIdAndOrder = VerticesIdAndOrder.assignVertexIds(dag);
         final int defaultParallelism = nodeEngine.getConfig().getJetConfig().getCooperativeThreadCount();
-        final Map<MemberInfo, int[]> partitionsByMember = getPartitionAssignment(nodeEngine, memberInfos);
-        final Map<Address, int[]> partitionsByAddress =
+        Map<MemberInfo, int[]> partitionsByMember = getPartitionAssignment(nodeEngine, memberInfos);
+        Map<Address, int[]> partitionsByAddress =
                 partitionsByMember.entrySet().stream().collect(toMap(en -> en.getKey().getAddress(), Entry::getValue));
-        final List<Address> addresses = toList(partitionsByMember.keySet(), MemberInfo::getAddress);
-        final int clusterSize = partitionsByMember.size();
+        int clusterSize = partitionsByMember.size();
         final boolean isJobDistributed = clusterSize > 1;
         final EdgeConfig defaultEdgeConfig = nodeEngine.getConfig().getJetConfig().getDefaultEdgeConfig();
         final Map<MemberInfo, ExecutionPlan> plans = new HashMap<>();
         int memberIndex = 0;
+        Set<Integer> requiredPartitions = jobConfig.getArgument(KEY_REQUIRED_PARTITIONS);
+        if (requiredPartitions != null) {
+            Map<MemberInfo, int[]> requiredMembers = new HashMap<>();
+            for (Entry<MemberInfo, int[]> entry : partitionsByMember.entrySet()) {
+                for (int partitionId : entry.getValue()) {
+                    if (requireNonNull(requiredPartitions).contains(partitionId)) {
+                        requiredMembers.putIfAbsent(entry.getKey(), entry.getValue());
```

Comment:
Here we keep all partitions owned by the member that owns at least one required partition, right?

What ensures that the dag doesn't contain a step which needs a particular member, e.g. for keyed aggregation operation?

Reply:
@frant-hartm it would be analyzed on SQL opt phase, so, SQL guarantees correctness.